onnx/defs/schema.h

```diff
@@ -33,6 +33,16 @@ typedef std::unordered_set<DataType> DataTypeSet;
 typedef std::unordered_map<std::string, std::pair<DataTypeSet, std::string>>
     TypeConstraintMap;
 
+typedef TensorShapeProto_Dimension InferenceDimension;
+
+struct InferenceContext {
+  virtual const AttributeProto* getAttribute(std::string name) = 0;
+  virtual const std::vector<const TypeProto_Tensor *>& getAllInputTypes() = 0;
+  virtual void setAllOutputTypes(std::vector<TypeProto_Tensor>&& allOutputTypes) = 0;
```

Comment:
personal suggestion:
change virtual void setAllOutputTypes(std::vector<TypeProto_Tensor>&& allOutputTypes) = 0; to 
virtual std::vector<const TypeProto_Tensor *>& getAllOutputTypes() = 0; or
virtual std::vector<const TypeProto_Tensor *>* getAllOutputTypes() = 0; if folks have concern on using non-const reference :)

Reply:
I'll go with the latter one - the problem with returning pointers is that the values may live on the stack, so they have to be copied out

Reply:
to be clear, setAllOutputTypes is used by a shape inference function to tell the InferenceContext what the inferred types are.

the problem with the first suggestion (`virtual std::vector<const TypeProto_Tensor *>& getAllOutputTypes() = 0;`) is that the `TypeProto_Tensor`s have no owner - they probably lived on the stack inside the shape inference function, which is now returning.

the problem with the second suggestion is that only allows reading the values, not setting them.

I might have a neater solution though